### PR TITLE
[MONDRIAN-2670] Regular Expression used in Analyzer's Find dialog not…

### PR DESCRIPTION
**Pull request by bennychow**
… passed down to query when using Vertica

### CODE REVIEW

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 8m
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl mondrian
```
---
 :exclamation: No tests found!
 ---
 
 

:information_source:  _This is an automatic message_

**Comment by bennychow** [comment]
@pentaho/millenniumfalcon @lucboudreau @mkambol @kurtwalker Please help review and merge.  Thanks!

File: mondrian/src/main/java/mondrian/spi/impl/VerticaDialect.java [hunk 1]
```diff
@@ -27,108 +31,165 @@
...
-
-    /**
-     * Creates a VerticaDialect.
-     *
-     * @param connection Connection
-     */
-    public VerticaDialect(Connection connection) throws SQLException {
-        super(connection);
-    }
+  private final String flagsRegexp = "^(\\(\\?([a-zA-Z]+)\\)).*$";
```
Comment (mkambol) [inline mondrian/src/main/java/mondrian/spi/impl/VerticaDialect.java]: It looks like this regex for finding the flags gets duplicated in other dialects.  Could the flag logic be extracted and moved either into JdbcDialectImpl or some utility class?

**Comment by mkambol** [comment]
Looks good @bennychow.  Just one small comment about potentially extracting some duplicated logic.

File: mondrian/src/main/java/mondrian/spi/impl/VerticaDialect.java (hunk 1 above)
Comment (bennychow) [reply mondrian/src/main/java/mondrian/spi/impl/VerticaDialect.java]: Thanks for reviewing Matt.  

It sure can.  I'm going to be working on getting regexp support for Redshift next so now is as good as time as ever to re-factor this a bit.

Comment (lucboudreau) [reply mondrian/src/main/java/mondrian/spi/impl/VerticaDialect.java]: Hey @bennychow I had this around my local box. I did some tests for Redshift and regular expressions. Hope this helps.

```
    @Override
    public boolean allowsRegularExpressionInWhereClause() {
        return true;
    }

    @Override
    public String generateRegularExpression(
        String source,
        String javaRegex)
    {
    	try {
            Pattern.compile(javaRegex);
        } catch (PatternSyntaxException e) {
            // Not a valid Java regex. Too risky to continue.
            return null;
        }

        // We might have to use case-insensitive matching
        javaRegex = DialectUtil.cleanUnicodeAwareCaseFlag(javaRegex);
        final Matcher flagsMatcher = flagsPattern.matcher(javaRegex);
        boolean caseSensitive = true;
        if (flagsMatcher.matches()) {
            final String flags = flagsMatcher.group(2);
            if (flags.contains("i")) {
                caseSensitive = false;
            }
        }
        if (flagsMatcher.matches()) {
            javaRegex =
                javaRegex.substring(0, flagsMatcher.start(1))
                + javaRegex.substring(flagsMatcher.end(1));
        }
        final Matcher escapeMatcher = escapePattern.matcher(javaRegex);
        while (escapeMatcher.find()) {
            javaRegex =
                javaRegex.replace(
                    escapeMatcher.group(1),
                    escapeMatcher.group(2));
        }
        final StringBuilder sb = new StringBuilder();

        // Now build the string.
        sb.append(source);
        sb.append(" IS NOT NULL AND ");
        
        // https://docs.aws.amazon.com/redshift/latest/dg/REGEXP_INSTR.html
        sb.append("REGEXP_INSTR(");
        sb.append(source);
        sb.append(",");
        quoteStringLiteral(sb, javaRegex);
        sb.append(",1,1,");
        sb.append(caseSensitive ? "c" : "i");
        sb.append(") > 0");

        return sb.toString();
    }
```

Comment (bennychow) [reply mondrian/src/main/java/mondrian/spi/impl/VerticaDialect.java]: Thanks Luc!  This will definitely be helpful.

**Pull request closed without merging**